data/libs/EquipSet.lua

```diff
@@ -1,404 +1,643 @@
+-- centralizes the management of "capability properties" set on the owning ship.
 
 ---@class EquipSet
----@field meta table
-local EquipSet = utils.inherits(nil, "EquipSet")
-
-EquipSet.default = {
-	cargo=0,
-	engine=1,
-	laser_front=1,
-	laser_rear=0,
-	missile=0,
-	ecm=1,
-	radar=1,
-	target_scanner=1,
-	hypercloud=1,
-	hull_autorepair=1,
-	energy_booster=1,
-	atmo_shield=1,
-	cabin=50,
-	shield=9999,
-	scoop=2,
-	laser_cooler=1,
-	cargo_life_support=1,
-	autopilot=1,
-	trade_computer=1,
-	sensor = 2,
-	thruster = 1
-}
-
-function EquipSet.New (slots)
-	---@class EquipSet
-	local obj = {}
-	obj.slots = {}
-	for k, n in pairs(EquipSet.default) do
-		obj.slots[k] = {__occupied = 0, __limit = n}
-	end
-	for k, n in pairs(slots) do
-		obj.slots[k] = {__occupied = 0, __limit = n}
-	end
-	setmetatable(obj, EquipSet.meta)
-	return obj
+---@field New fun(ship: Ship): EquipSet
+local EquipSet = utils.class("EquipSet")
+
+---@alias EquipSet.Listener fun(op: 'install'|'remove', equip: EquipType, slot: HullConfig.Slot?)
+
+-- Function: SlotTypeMatches
+--
+-- Static helper function that performs filtering of slot type identifiers.
+--
+-- Call this function when you want to know if the actual type of some object
+-- is a valid match with the given filter string.
+--
+-- Example:
+--
+--  > local validEquipForSlot = EquipSet.SlotTypeMatches(equip.slot.type, slot.type)
+--
+-- Parameters:
+--
+--  actualType - string, concrete fully-qualified type string of the object or
+--               slot in question
+--
+--  filter     - string, the partially-qualified type to check the concrete type
+--               against to determine a match
+--
+---@param actualType string
+---@param filter string
+local function slotTypeMatches(actualType, filter)
+	return actualType == filter or string.sub(actualType, 1, #filter + 1) == filter .. "."
 end
 
-local listeners = {}
-function EquipSet:AddListener(listener)
-	listeners[self] = listeners[self] or {}
-	table.insert(listeners[self], listener)
+EquipSet.SlotTypeMatches = slotTypeMatches
+
+-- Function: CompatibleWithSlot
+--
+-- Static helper function to check if the given equipment item is compatible
+-- with the given slot object. Validates type and size parameters of the slot.
+--
+-- Parameters:
+--
+--  equip - EquipType, equipment item instance or prototype to check
+--
+--  slot  - optional HullConfig.Slot, the slot the equipment item is being
+--          validated against. If not present, the function validates that the
+--          passed equipment item is a non-slot equipment item.
+--
+---@param equip EquipType
+---@param slot HullConfig.Slot?
+function EquipSet.CompatibleWithSlot(equip, slot)
+	local equipSlot = equip.slot or false
+	if not slot then return not equipSlot end
+
+	return equipSlot and
+		slotTypeMatches(equip.slot.type, slot.type)
+		and (equip.slot.size <= slot.size)
+		and (equip.slot.size >= (slot.size_min or slot.size))
```

Comment:
Nitpick: since you've made the alias, should use it.  Alternatively drop the `local equipSlot`

```suggestion
		slotTypeMatches(equipSlot.type, slot.type)
		and (equipSlot.size <= slot.size)
		and (equipSlot.size >= (slot.size_min or slot.size))
```

Reply:
Good catch, thank you!